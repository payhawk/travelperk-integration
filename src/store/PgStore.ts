import * as fs from 'fs';

import * as moment from 'moment';
import { Pool } from 'pg';

import { ILogger } from '@utils';

import { SCHEMA } from './Config';
import { IInvoicesSyncHistoryItemRecord, INewUserTokenSetRecord, ISchemaStore, IUserTokenSetRecord, PayhawkApiKeyRecordKeys, SyncHistoryItemRecordKeys, UserTokenSetRecordKeys } from './contracts';

export class PgStore implements ISchemaStore {

    constructor(private readonly pgClient: Pool, private readonly logger: ILogger) {
    }

    async getAllTokenSets(): Promise<IUserTokenSetRecord[]> {
        const result = await this.pgClient.query<IUserTokenSetRecord>({
            text: `SELECT * FROM "${SCHEMA.TABLE_NAMES.ACCESS_TOKENS}"`,
        });

        return result.rows;
    }

    async saveAccessToken({ account_id, token_set }: INewUserTokenSetRecord): Promise<void> {
        await this.pgClient.query({
            text: `
                INSERT INTO "${SCHEMA.TABLE_NAMES.ACCESS_TOKENS}" (
                    "${UserTokenSetRecordKeys.account_id}",
                    "${UserTokenSetRecordKeys.token_set}"
                )
                VALUES ($1, $2)
                ON CONFLICT ("${UserTokenSetRecordKeys.account_id}")
                DO
                    UPDATE SET
                        "${UserTokenSetRecordKeys.token_set}" = $2,
                        "${UserTokenSetRecordKeys.updated_at}" = now();
            `,
            values: [
                account_id,
                token_set,
            ],
        });
    }

    async getAccessToken(accountId: string): Promise<IUserTokenSetRecord | undefined> {
        const query = await this.pgClient.query<IUserTokenSetRecord>({
            text: `
                SELECT * FROM "${SCHEMA.TABLE_NAMES.ACCESS_TOKENS}"
                WHERE "${UserTokenSetRecordKeys.account_id}"=$1
            `,
            values: [
                accountId,
            ],
        });

        const record = query.rows[0];
        return record;
    }

    async deleteAccessToken(accountId: string): Promise<void> {
        await this.pgClient.query<IUserTokenSetRecord>({
            text: `
                DELETE FROM "${SCHEMA.TABLE_NAMES.ACCESS_TOKENS}"
                WHERE "${UserTokenSetRecordKeys.account_id}"=$1
            `,
            values: [
                accountId,
            ],
        });
    }

    async getApiKey(accountId: string): Promise<string|undefined> {
        const query = await this.pgClient.query<{ key: string }>({
            text: `
                SELECT "${PayhawkApiKeyRecordKeys.key}" FROM "${SCHEMA.TABLE_NAMES.PAYHAWK_API_KEYS}"
                WHERE "${PayhawkApiKeyRecordKeys.account_id}" = $1
            `,
            values: [accountId],
        });

        if (query.rows.length > 0) {
            return query.rows[0].key;
        } else {
            return undefined;
        }
    }

    async setApiKey(accountId: string, key: string): Promise<void> {
        await this.pgClient.query<{ payhawk_api_key: string }>({
            text: `
                INSERT INTO "${SCHEMA.TABLE_NAMES.PAYHAWK_API_KEYS}" ("${PayhawkApiKeyRecordKeys.account_id}", "${PayhawkApiKeyRecordKeys.key}")
                VALUES ($1, $2)
                ON CONFLICT ("${PayhawkApiKeyRecordKeys.account_id}")
                DO
                    UPDATE SET "${PayhawkApiKeyRecordKeys.key}" = $2, "${PayhawkApiKeyRecordKeys.updated_at}" = NOW()
            `,
            values: [accountId, key],
        });
    }

    async getLastSyncDate(accountId: string): Promise<Date | undefined> {
        const query = await this.pgClient.query<Pick<IInvoicesSyncHistoryItemRecord, 'last_sync_at'>>({
            text: `
                SELECT "${SyncHistoryItemRecordKeys.last_sync_at}" FROM "${SCHEMA.TABLE_NAMES.INVOICES_SYNC_HISTORY}"
                WHERE "${PayhawkApiKeyRecordKeys.account_id}" = $1
            `,
            values: [accountId],
        });

        if (query.rows.length > 0) {
            return moment.utc(query.rows[0].last_sync_at).toDate();
        }

        return undefined;
    }

    async updateLastSyncDate(accountId: string, lastSyncAt: Date): Promise<void> {
        await this.pgClient.query({
            text: `
                INSERT INTO "${SCHEMA.TABLE_NAMES.INVOICES_SYNC_HISTORY}"
                    ("${SyncHistoryItemRecordKeys.account_id}", "${SyncHistoryItemRecordKeys.last_sync_at}")
                VALUES ($1, $2)
                ON CONFLICT ("${SyncHistoryItemRecordKeys.account_id}")
                DO
                    UPDATE SET "${SyncHistoryItemRecordKeys.last_sync_at}" = $2
            `,
            values: [accountId, lastSyncAt],
        });
    }

    async initSchema(): Promise<void> {
        await this.pgClient.query(await readSchemaInitScript());
    }

    async ensureSchemaVersion(): Promise<void> {
        await this.applyMigration();
    }

    private async applyMigration(): Promise<void> {
        await this.applyDatabaseMigration();
    }

    private async applyDatabaseMigration(): Promise<void> {
        const fileName = `migration.sql`;
        if (!scriptExists(fileName)) {
            this.logger.info('Database migration skipped. No script');
            return;
        }

        try {
            this.logger.info('Database migration started');

            const scriptText = await readScript(fileName);
            await this.pgClient.query(scriptText);

            this.logger.info('Database migration finished');
        } catch (err) {
            const error = Error(`Database migration script failed: ${err.toString()}`);

            this.logger.error(error);
        }
    }
}

const readSchemaInitScript = async (): Promise<string> => {
    return readScript('init.schema.sql');
};

const readScript = async (name: string): Promise<string> => {
    return await new Promise<string>((resolve, reject) => {
        fs.readFile(getPathFullName(name), 'utf8', (err, data) => {
            err ? reject(err) : resolve(data);
        });
    });
};

const scriptExists = (name: string): boolean => {
    return fs.existsSync(getPathFullName(name));
};

const getPathFullName = (fileName: string): string => {
    return `${process.cwd()}/assets/${fileName}`;
};
