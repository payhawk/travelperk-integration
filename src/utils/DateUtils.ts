import * as moment from 'moment';

export const toShortDateFormat = (date: moment.Moment | Date | string): string => {
    return moment.utc(date).format('YYYY-MM-DD');
};

export const toShortDate = (date: moment.Moment | Date | string): Date => {
    return moment.utc(date).hours(0).minutes(0).seconds(0).toDate();
};
