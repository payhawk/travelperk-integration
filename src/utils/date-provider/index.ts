import { IDateProvider } from './IDateProvider';
import { MomentDateProvider } from './MomentDateProvider';

export * from './IDateProvider';

export type IDateProviderFactory = () => IDateProvider;

export const createDateProvider: IDateProviderFactory = () => new MomentDateProvider();
