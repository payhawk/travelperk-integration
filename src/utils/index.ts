export * from './logger';
export * from './request';
export * from './date-provider';

export * from './Base64Converter';
export * from './DateUtils';
export * from './CommonTypes';
