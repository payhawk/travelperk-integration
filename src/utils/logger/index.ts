import * as pino from 'pino';

import { config } from '@config';

import { ILogger } from './ILogger';
import { PinoStackDriverLogger } from './PinoStackDriverLogger';

export * from './ILogger';

let logger: ILogger;

export const createLogger = (): ILogger => {
    if (!logger) {
        logger = new PinoStackDriverLogger(config.serviceName, pino());
    }

    return logger;
};
