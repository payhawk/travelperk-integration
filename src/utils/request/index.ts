export * from './RequestValidation';
export * from './RequestHandler';
