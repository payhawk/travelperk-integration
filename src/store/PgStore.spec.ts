import { Pool } from 'pg';
import { It, Mock } from 'typemoq';

import { ILogger } from '../utils';
import { PgStore } from './PgStore';

describe('PgStore', () => {
    const dbClientMock = Mock.ofType<Pool>();
    const loggerMock = Mock.ofType<ILogger>();

    const store = new PgStore(dbClientMock.object, loggerMock.object);

    afterEach(() => {
        [
            dbClientMock,
            loggerMock,
        ].forEach(m => {
            m.verifyAll();
            m.reset();
        });
    });

    it('should match snapshot for getting all tokens', async () => {
        dbClientMock
            .setup(d => d.query(It.isAny()))
            .callback(req => expect(req).toMatchSnapshot())
            .returns(async () => ({ rows: [], rowCount: 0 }));

        await store.getAllTokenSets();
    });

    it('should match snapshot for getting api key', async () => {
        dbClientMock
            .setup(d => d.query(It.isAny()))
            .callback(req => expect(req).toMatchSnapshot())
            .returns(async () => ({ rows: [], rowCount: 0 }));

        await store.getApiKey('acc_id');
    });

    it('should match snapshot for setting api key', async () => {
        dbClientMock
            .setup(d => d.query(It.isAny()))
            .callback(req => expect(req).toMatchSnapshot())
            .returns(async () => ({ rows: [], rowCount: 0 }));

        await store.setApiKey('acc_id', 'key');
    });

    it('should match snapshot for saving access token', async () => {
        dbClientMock
            .setup(d => d.query(It.isAny()))
            .callback(req => expect(req).toMatchSnapshot())
            .returns(async () => ({ rows: [], rowCount: 0 }));

        await store.saveAccessToken({
            account_id: 'acc_id',
            token_set: { access_token: 'a', refresh_token: 'r', scope: 's', token_type: 'Bearer' },
        });
    });

    it('should match snapshot for deleting access token', async () => {
        dbClientMock
            .setup(d => d.query(It.isAny()))
            .callback(req => expect(req).toMatchSnapshot())
            .returns(async () => ({ rows: [], rowCount: 0 }));

        await store.deleteAccessToken('acc_id');
    });

    it('should match snapshot for getting access token for account', async () => {
        dbClientMock
            .setup(d => d.query(It.isAny()))
            .callback(req => expect(req).toMatchSnapshot())
            .returns(async () => ({ rows: [], rowCount: 0 }));

        await store.getAccessToken('acc_id');
    });

    it('should match snapshot for getting last sync date for account', async () => {
        dbClientMock
            .setup(d => d.query(It.isAny()))
            .callback(req => expect(req).toMatchSnapshot())
            .returns(async () => ({ rows: [], rowCount: 0 }));

        await store.getLastSyncDate('acc_id');
    });

    it('should match snapshot for getting last sync date for account', async () => {
        dbClientMock
            .setup(d => d.query(It.isAny()))
            .callback(req => expect(req).toMatchSnapshot())
            .returns(async () => ({ rows: [], rowCount: 0 }));

        await store.updateLastSyncDate('acc_id', new Date(Date.UTC(2020, 1, 2)));
    });
});
