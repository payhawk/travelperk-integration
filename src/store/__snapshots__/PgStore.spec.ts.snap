// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PgStore should match snapshot for deleting access token 1`] = `
Object {
  "text": "
                DELETE FROM \\"access_tokens\\"
                WHERE \\"account_id\\"=$1
            ",
  "values": Array [
    "acc_id",
  ],
}
`;

exports[`PgStore should match snapshot for getting access token for account 1`] = `
Object {
  "text": "
                SELECT * FROM \\"access_tokens\\"
                WHERE \\"account_id\\"=$1
            ",
  "values": Array [
    "acc_id",
  ],
}
`;

exports[`PgStore should match snapshot for getting all tokens 1`] = `
Object {
  "text": "SELECT * FROM \\"access_tokens\\"",
}
`;

exports[`PgStore should match snapshot for getting api key 1`] = `
Object {
  "text": "
                SELECT \\"key\\" FROM \\"payhawk_api_keys\\"
                WHERE \\"account_id\\" = $1
            ",
  "values": Array [
    "acc_id",
  ],
}
`;

exports[`PgStore should match snapshot for getting last sync date for account 1`] = `
Object {
  "text": "
                SELECT \\"last_sync_at\\" FROM \\"invoices_sync_history\\"
                WHERE \\"account_id\\" = $1
            ",
  "values": Array [
    "acc_id",
  ],
}
`;

exports[`PgStore should match snapshot for getting last sync date for account 2`] = `
Object {
  "text": "
                INSERT INTO \\"invoices_sync_history\\"
                    (\\"account_id\\", \\"last_sync_at\\")
                VALUES ($1, $2)
                ON CONFLICT (\\"account_id\\")
                DO
                    UPDATE SET \\"last_sync_at\\" = $2
            ",
  "values": Array [
    "acc_id",
    2020-02-02T00:00:00.000Z,
  ],
}
`;

exports[`PgStore should match snapshot for saving access token 1`] = `
Object {
  "text": "
                INSERT INTO \\"access_tokens\\" (
                    \\"account_id\\",
                    \\"token_set\\"
                )
                VALUES ($1, $2)
                ON CONFLICT (\\"account_id\\")
                DO
                    UPDATE SET
                        \\"token_set\\" = $2,
                        \\"updated_at\\" = now();
            ",
  "values": Array [
    "acc_id",
    Object {
      "access_token": "a",
      "refresh_token": "r",
      "scope": "s",
      "token_type": "Bearer",
    },
  ],
}
`;

exports[`PgStore should match snapshot for setting api key 1`] = `
Object {
  "text": "
                INSERT INTO \\"payhawk_api_keys\\" (\\"account_id\\", \\"key\\")
                VALUES ($1, $2)
                ON CONFLICT (\\"account_id\\")
                DO
                    UPDATE SET \\"key\\" = $2, \\"updated_at\\" = NOW()
            ",
  "values": Array [
    "acc_id",
    "key",
  ],
}
`;
